add mkldnn linear backward

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* #37146 add mkldnn pooling backward
* **#36122 add mkldnn linear backward**
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440966](https://our.internmc.facebook.com/intern/diff/D22440966)

XiaobingSuper committed:
add mkldnn linear backward

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/mkldnn/Linear.cpp:
```diff
@@ -32,17 +48,16 @@ Tensor mkldnn_linear(
       "mkldnn_linear: input needs to has dim at least 2, input dim ", self.dim());
   TORCH_CHECK(self.is_mkldnn(),
       "mkldnn_linear: input needs to be mkldnn layout");
-  TORCH_CHECK(weight.is_mkldnn() && bias.is_mkldnn(),
```
The backward formulas seems to actually assume that the weights are no mkldnn Tensors (because mkldnn_linear_backward_weights always convert to dense() the outputs before returning them). Is that true? If not, the backward formula should be updated to make sure it returns the same type as the weights.

XiaobingSuper replied:
For backward path, we want to support two method.

1. input is a MKLDNN tensor, weight is  a dense tensor, this our primary method, but this method always has weight reorder at each iteration.
2.  input is a MKLDNN tensor, weight is a MKLDNN tensor by call **model.to_mkldnn()** to transfer weight to a MKLDNN tensor before training,  \this method will reduce the weight reorder time, but for the optimizer step, we may meet some ops not support MKLDNN tensor, and also have a problem when we save the model: t**orch.sav**e can't save MKLDNN tensor, we may can solve those problems to get more better performance in the future .

albanD replied:
Ok, So the backward formula needs to be updated to properly support case 2 right?

XiaobingSuper replied:
Sorry for confuse you, for the first step, backward formula support case 1, I remove this check because: for inference, input , weight and bias are MKLDNN tensor, for training, input is MKLDNN tensor, weight  and bias are dense tensor.

albanD replied:
Sounds good.
Could you add the corresponding checks for the backward function to be sure that you have what you want. Nothing else prevent the user from doing training with weight/bias as MKLDNN Tensors right?

XiaobingSuper replied:
Add some checks for backward function.

XiaobingSuper commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3318,6 +3327,7 @@
     CUDA: legacy::cuda::_th_addmm_out
     SparseCPU: addmm_out_sparse_dense_cpu
     SparseCUDA: addmm_out_sparse_dense_cuda
+    MkldnnCPU: mkldnn_linear_wraper_out
```
Yes, I will change it.

XiaobingSuper commented on tools/autograd/derivatives.yaml:
```diff
@@ -1572,6 +1570,9 @@
 - name: mkldnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
   grad_output, self, weight: _convolution_double_backward(grads[0], grads[1], grads[2], grad_output, weight, self, stride, padding, dilation, false, std::vector<int64_t>(padding.size(), 0), groups, false, false, false, grad_input_mask)
 
+- name: mkldnn_linear(Tensor self, Tensor weight, Tensor? bias=None) -> Tensor
```
This is not used now, but when we port aten::linear, this will be needed.

albanD replied:
Could we keep this out until then? Every entry here adds a lot of generated code. So if we can keep it low, it's better :D

XiaobingSuper replied:
Removed.

XiaobingSuper commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -572,6 +572,25 @@ Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntArrayRef si
+std::tuple<Tensor, Tensor, Tensor> addmm_backward(const Tensor& grad_output, const Tensor& mat1, const Tensor& mat2,
+    const Scalar& beta, const Scalar& alpha, std::array<bool, 3> output_mask) {
+  Tensor grad1, grad2, grad3;
+  if (grad_output.is_mkldnn()) {
```
Yes, you are right.

XiaobingSuper commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -12,6 +12,22 @@
+    const Tensor& input, const Tensor& weight, Scalar beta, Scalar alpha) {
+  TORCH_CHECK(input.dim() == 2,
+      "mkldnn_linear__wraper_out: input needs to has dim 2, input dim ", input.dim());
+  result = at::mkldnn_linear(input, weight.t(), bias);
```
Because there only have  **mkldnn_linear** API, if it is inappropriate, I need add a new API named mkldnn_linear_out to support it.

XiaobingSuper replied:
Now, will give a warning that MKLDNN not support it.

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1655,11 +1655,20 @@
   python_module: nn
 
-- func: mkldnn_linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor
+- func: mkldnn_linear(Tensor self, Tensor weight, Tensor? bias=None) -> Tensor
```
keep the old name now.

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

albanD reviewed (approved):
lgtm. Thanks for updating.

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"


Differential Revision: [D22440966](https://our.internmc.facebook.com/intern/diff/D22440966)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn linear backward"


Differential Revision: [D22440966](https://our.internmc.facebook.com/intern/diff/D22440966)

[ghstack-poisoned]

ZhuJewel commented:
> lgtm. Thanks for updating.

@albanD , may I know if the PR can be merged? Thanks

albanD commented:
Hey,

Yes this is in the process.
We have some timeout issues on intern CI and with such a big stack it delays the land a bit. But it's on the way!

ZhuJewel commented:
Hey, @albanD , anything we can help in our side for the CI errors, please feel free to let us know.
